benchmarks/experimental_components/examples/end_to_end.py

```diff
@@ -0,0 +1,291 @@
+        return sorted(x.items(), key=lambda i: (i[1], i[0]))
+    ordered_descriptions = [i[0] for i in sort_by_value(relative_improvement)]
+
+    n_regressed = len([i for i in relative_improvement.values() if i < -0.1])
+    n_improved = len([i for i in relative_improvement.values() if i > 0.1])
+    n_similar = len(relative_improvement) - n_improved - n_regressed
+    print()
+    for label, n in [
+        ("Improved  (>10%):", n_improved),
+        ("Regressed (>10%):", n_regressed),
+        ("Within 10%:", n_similar),
+    ]:
+        print(f"{label:<17} {n:>4}  ({n / len(relative_improvement) * 100:.0f}%)")
+
+    print(f"\nImprovement    Absolute  |{'':>7}dtype{'':>6}numel   ", end="")
+    print(f"mask_reuse     mask_true_pct      x_layout    mask_layout\n{'=' * 114}")
+
+    def print_block(result_filter):
+        print_result_block(
+            master_results + branch_results, result_filter,
+            relative_improvement, absolute_times,
+        )
+    print_block(ordered_descriptions[:15])
+    print("\n...\n")
+    print_block(ordered_descriptions[-15:])
+
+
+def print_result_block(
+    full_results, result_filter, relative_improvement, absolute_times
+):
+    results = [i for i in full_results if i.description in result_filter]
+    labels = {r.description: pretty_str(r) for r in results}
+    labels = sorted(labels.items(), key=lambda i: relative_improvement[i[0]])
+
+    for key, value in labels:
+        delta_t = absolute_times[_MASTER][key] - absolute_times[_BRANCH][key]
+        print(
+            f" {relative_improvement[key] * 100:>5.0f}%   "
+            f"{abs(delta_t) * 1e6:>10.1f} us  |   "
+            f"{key.split(']')[1][2:-1]:>9}  {value}"
+        )
+
+
+def pretty_str(m):
+    tensor_parameters = m.metadata["tensor_parameters"]
+    x_numel = tensor_parameters["x"]["numel"]
+    mask_numel = tensor_parameters["mask"]["numel"]
+    mask_reuse = int(x_numel / mask_numel)
+
+    def order(name):
+        return str("contiguous" if tensor_parameters[name]["is_contiguous"]
+                   else tensor_parameters[name]["order"])
+    return (
+        f"{x_numel:>9}  {mask_reuse:>9}{'':>12}"
+        f"{m.metadata['params']['mask_fraction'] * 100:>3.0f}%      "
+        f"{order('x'):>12}   {order('mask'):>12}"
+    )
+
+
+# \\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\
+# == Subprocess and environment management ====================================
+# /////////////////////////////////////////////////////////////////////////////
+def read_results(result_file: str):
+    output = []
+    try:
+        with open(result_file, "rb") as f:
+            while True:
+                try:
+                    output.append(pickle.load(f))
+                except EOFError:
+                    break
+    finally:
+        os.remove(result_file)
+    return output
+
+
+def invoke_subprocess(env: str, seed: int):
+    _, result_file = tempfile.mkstemp(suffix=".pkl")
+    subprocess.run(
```

Comment:
why subprocess and not just use multiprocessing? (it takes care of launching the right interpreter and loading the right modules)

Reply:
Two reasons:
1) For this example, we're actually using two different conda environments and two different pytorch builds, so the sort of forking / state sharing that multiprocessing does is not appropriate.

2) I have a deep and enduring suspicion of multiprocessing after one too many maddening deadlocks. I hear moving from `fork` to `spawn` makes things better, but once bitten twice shy. ^^